src/nbl/video/COpenGLCommandBuffer.cpp

```diff
@@ -0,0 +1,743 @@
+        {
+            // TODO: check it->bufferOffset is aligned to data type of E_FORMAT
+            //assert(?);
+
+            uint32_t pitch = ((it->bufferRowLength ? it->bufferRowLength : it->imageExtent.width) * bpp).getIntegerApprox();
+            int32_t alignment = 0x1 << core::min(core::max(core::findLSB(it->bufferOffset), core::findLSB(pitch)), 3u);
+            ctxlocal->nextState.pixelPack.alignment = alignment;
+            ctxlocal->nextState.pixelPack.rowLength = it->bufferRowLength;
+            ctxlocal->nextState.pixelPack.imgHeight = it->bufferImageHeight;
+
+            auto yStart = type == IGPUImage::ET_1D ? it->imageSubresource.baseArrayLayer : it->imageOffset.y;
+            auto yRange = type == IGPUImage::ET_1D ? it->imageSubresource.layerCount : it->imageExtent.height;
+            auto zStart = type == IGPUImage::ET_2D ? it->imageSubresource.baseArrayLayer : it->imageOffset.z;
+            auto zRange = type == IGPUImage::ET_2D ? it->imageSubresource.layerCount : it->imageExtent.depth;
+            if (compressed)
+            {
+                ctxlocal->nextState.pixelPack.BCwidth = blockDims[0];
+                ctxlocal->nextState.pixelPack.BCheight = blockDims[1];
+                ctxlocal->nextState.pixelPack.BCdepth = blockDims[2];
+
+                ctxlocal->flushStateGraphics(gl, SOpenGLContextLocalCache::GSB_PIXEL_PACK_UNPACK, ctxid);
+
+                // TODO impl in func table
+                //gl->extGlGetCompressedTextureSubImage(src, it->imageSubresource.mipLevel, it->imageOffset.x, yStart, zStart, it->imageExtent.width, yRange, zRange,
+                //    dstBuffer->getSize() - it->bufferOffset, reinterpret_cast<void*>(it->bufferOffset));
+            }
+            else
+            {
+                ctxlocal->flushStateGraphics(gl, SOpenGLContextLocalCache::GSB_PIXEL_PACK_UNPACK, ctxid);
+
+                // TODO impl in func table
+                //gl->extGlGetTextureSubImage(src, it->imageSubresource.mipLevel, it->imageOffset.x, yStart, zStart, it->imageExtent.width, yRange, zRange,
+                //    glfmt, gltype, dstBuffer->getSize() - it->bufferOffset, reinterpret_cast<void*>(it->bufferOffset));
+            }
+        }
+    }
+
+    void COpenGLCommandBuffer::beginRenderpass_clearAttachments(IOpenGL_FunctionTable* gl, const SRenderpassBeginInfo& info, GLuint fbo)
+    {
+        auto& rp = info.framebuffer->getCreationParameters().renderpass;
+        auto& sub = rp->getSubpasses().begin()[0];
+        auto* color = sub.colorAttachments;
+        auto* depthstencil = sub.depthStencilAttachment;
+        auto* descriptions = rp->getAttachments().begin();
+
+        for (uint32_t i = 0u; i < sub.colorAttachmentCount; ++i)
+        {
+            // TODO how do i set clear color in vulkan ???
+            GLfloat colorf[4]{ 0.f, 0.f, 0.f, 0.f };
+            GLint colori[4]{ 0,0,0,0 };
+            GLuint coloru[4]{ 0u,0u,0u,0u };
```

Comment:
https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkClearValue.html

> This union is used where part of the API requires either color or depth/stencil clear values, depending on the attachment, and defines the initial clear values in the VkRenderPassBeginInfo structure.

Reply:
yeah, but renderpass defines what happens at the beginning with attachments... and i thought if it defines CLEAR behaviour, then an attachment is cleared automatically. Turns out i was wrong and you to do the clears manually anyway...

Reply:
yep, its part of the begin renderpass command for the commandbuffer